### ISSUE TITLE
Psalm throws an error if class property set by public function called from constructor

### ISSUE THREAD

**Issue by nickyr** [vimeo/psalm#92]
Paste the following into the box on getpsalm.org.

```php
<?php
class A
{
    /**
    * @var int
    */
    protected $p;
  
    public function __construct()
    {
        $this->initialize();   	 
    }
  
    public function initialize()
    {
     	$this->p = 3; 
    }
}
```

and you get

![image](https://cloud.githubusercontent.com/assets/1389687/22804486/43076942-eee7-11e6-9ab3-40add6192d1d.png)

Why can't it be set in a public function?

**Comment by muglug** [comment]
Because public/protected functions can be overridden e.g.

```php
class A
{
    /**
     * @var int
     */
    protected $p;
  
    public function __construct()
    {
        $this->initialize();   	 
    }
  
    public function initialize()
    {
     	$this->p = 3; 
    }
}

class B extends A {
    public function initialize()
    {
     	// do nothing
    }
}
```

**Comment by nickyr** [comment]
well it shouldn't be an error unless it is overridden and doesn't set the variable there, right?

**Comment by muglug** [comment]
Oh, this could totally be a false positive, and you're welcome to ignore the issue in your config.

But with this policy, I'm opting for cautiousness over callousness.

Also, Hack has an identical behaviour - try pasting in
```php
<?hh
class A
{
    private int $p;
  
    public function __construct() {
        $this->initialize();
    }
  
    public final function initialize() {
    	$this->p = 5;
    }
}
```

to the code box on http://hacklang.org/
